scripts/valhalla_build_extract.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+
+import argparse
+import ctypes
+from io import BytesIO
+import json
+import logging
+import os
+from pathlib import Path
+import struct
+import sys
+import tarfile
+from tarfile import BLOCKSIZE
+from time import time
+from typing import List, Tuple
+
+# "<" prefix means little-endian and no alignment
+# order is important! if uint64_t is not first, c++ will use padding bytes to unpack
+INDEX_BIN_FORMAT = '<QLL'
+INDEX_BIN_SIZE = struct.calcsize(INDEX_BIN_FORMAT)
+INDEX_FILE = "index.bin"
+# skip the first 40 bytes of the tile header
+GRAPHTILE_SKIP_BYTES = struct.calcsize('<Q2f16cQ')
+TRAFFIC_HEADER_SIZE = struct.calcsize('<2Q4I')
+TRAFFIC_SPEED_SIZE = struct.calcsize('<Q')
+
+
+class TileHeader(ctypes.Structure):
+    """
+    Resembles the uint64_t bit field at bytes 40 - 48 of the
+    graphtileheader to get the directededgecount_.
+    """
+    _fields_ = [
+        ("nodecount_", ctypes.c_ulonglong, 21),
+        ("directededgecount_", ctypes.c_ulonglong, 21),
+        ("predictedspeeds_count_", ctypes.c_ulonglong, 21),
+        ("spare1_", ctypes.c_ulonglong, 1)
+    ]
+
+
+description = "Builds a tar extract from the tiles in mjolnir.tile_dir to the path specified in mjolnir.tile_extract."
+parser = argparse.ArgumentParser(description=description)
+parser.add_argument("-c", "--config", help="Absolute or relative path to the Valhalla config JSON.", type=Path)
+parser.add_argument("-t", "--with-traffic", help="Flag to add a traffic.tar skeleton", action="store_true", default=False)
+parser.add_argument("-v", "--verbosity", help="Accumulative verbosity flags; -v: INFO, -vv: DEBUG", action='count', default=0)
+
+# set up the logger basics
+LOGGER = logging.getLogger(__name__)
+handler = logging.StreamHandler()
+handler.setFormatter(logging.Formatter("%(asctime)s %(levelname)5s: %(message)s"))
+LOGGER.addHandler(handler)
+
+
+def get_tile_count(in_path: Path) -> int:
+    """Recursively fills the passed map with tile path & size info"""
+    count = 0
+    for _, _, files in os.walk(in_path):
+        count += len(list(filter(lambda f: f.endswith('.gph'), files)))
+
+    return count
+
+
+def get_tile_id(path: str):
+    """Turns a tile path into a numeric GraphId"""
+    level, idx = path[:-4].split('/', 1)
+
+    return int(level) | (int(idx.replace('/', '')) << 3)
+
+
+def get_tar_info(name: str, size: int) -> tarfile.TarInfo:
+    """Creates and returns a tarinfo object"""
+    tarinfo = tarfile.TarInfo(name)
+    tarinfo.size = size
+    tarinfo.mtime = int(time())
+    tarinfo.type = tarfile.REGTYPE
+
+    return tarinfo
+
+
+def write_index_to_tar(tar_fp_: Path):
+    """Loop through all tiles and write the correct index.bin file to the tar"""
+    # get the offset and size from the tarred tile members
+    index: List[Tuple[int, int, int]] = list()
+    with tarfile.open(tar_fp_, 'r|') as tar:
```

Comment:
On second look: get rid of the write further down by using tarfile‘s fileobj in this loop which is already an open file descriptor to the tar file. so we don’t need to keep the index in memory, I think we can write it dynamically while parsing the members.

Reply:
well that didn't pan out, guess it's getting confused that its fileobj is changing underneath while it's traversing the files inside the tar..